File: loadercomplex-fabric/src/ver171/java/com/enderzombi102/loadercomplex/fabric17/impl/entity/FabricLivingEntity.java
package com.enderzombi102.loadercomplex.fabric17.impl.entity;

import com.enderzombi102.loadercomplex.api.minecraft.entity.PlayerEntity;
import com.enderzombi102.loadercomplex.fabric17.impl.item.FabricItemStack;
import com.enderzombi102.loadercomplex.api.minecraft.entity.LivingEntity;
import com.enderzombi102.loadercomplex.api.minecraft.item.EquipmentSlot;
import com.enderzombi102.loadercomplex.api.minecraft.item.ItemStack;
import net.minecraft.util.Hand;

public class FabricLivingEntity extends FabricEntity implements LivingEntity {
	private final net.minecraft.entity.LivingEntity backingEntity;

	public FabricLivingEntity(net.minecraft.entity.LivingEntity backingEntity) {
		super(backingEntity);
		this.backingEntity = backingEntity;
	}

	@Override
	public float getHealth() {
		return this.backingEntity.getHealth();
	}

	@Override
	public void setHealth(float health) {
		this.backingEntity.setHealth( health );
	}

	@Override
	public boolean isPlayer() {
		return this.backingEntity instanceof net.minecraft.entity.player.PlayerEntity;
	}

	@Override
	public PlayerEntity asPlayer() {
		return new FabricPlayer( (net.minecraft.entity.player.PlayerEntity) this.backingEntity );
	}

	@Override
	public boolean isChild() {
		return this.backingEntity.isBaby();
	}

	@Override
	public boolean isWaterMob() {
		return this.backingEntity.canBreatheInWater();
	}

	@Override
	public boolean isClimbing() {
		return this.backingEntity.isClimbing();
	}

	@Override
	public int getArmorProtection() {
		return this.backingEntity.getArmor();
	}

	@Override
	public ItemStack getItemInMainHand() {
		return new FabricItemStack( this.backingEntity.getStackInHand(Hand.MAIN_HAND) );
	}

	@Override
	public ItemStack getItemInOffHand() {
		return new FabricItemStack( this.backingEntity.getStackInHand(Hand.OFF_HAND) );
	}

	@Override
	public boolean hasEquipment( EquipmentSlot slot) {
		return this.backingEntity.hasStackEquipped( net.minecraft.entity.EquipmentSlot.valueOf( slot.name() ) );
	}

	@Override
	public ItemStack getEquipment( EquipmentSlot slot) {
		return new FabricItemStack( this.backingEntity.getEquippedStack( net.minecraft.entity.EquipmentSlot.valueOf( slot.name() ) ) );
	}

	@Override
	public void setEquipment( EquipmentSlot slot, ItemStack stack) {
		// TODO: Find this
	}
}

File: loadercomplex-fabric/src/ver171/kotlin/com/enderzombi102/loadercomplex/fabric171/impl/entity/FabricLivingEntity.kt
package com.enderzombi102.loadercomplex.fabric171.impl.entity

import com.enderzombi102.loadercomplex.api.minecraft.entity.LivingEntity
import com.enderzombi102.loadercomplex.api.minecraft.entity.PlayerEntity
import com.enderzombi102.loadercomplex.api.minecraft.item.EquipmentSlot
import com.enderzombi102.loadercomplex.api.minecraft.item.ItemStack
import com.enderzombi102.loadercomplex.fabric171.impl.item.FabricItemStack
import com.enderzombi102.loadercomplex.fabric171.impl.utils.toMC
import net.minecraft.entity.LivingEntity as McLivingEntity
import net.minecraft.entity.player.PlayerEntity as McPlayerEntity
import net.minecraft.util.Hand as McHand

open class FabricLivingEntity( private val backingEntity: McLivingEntity ) : FabricEntity( backingEntity ), LivingEntity {
	override fun getHealth(): Float =
		backingEntity.health

	override fun setHealth( health: Float ) {
		backingEntity.health = health
	}

	override fun isPlayer(): Boolean =
		backingEntity is McPlayerEntity

	override fun asPlayer(): PlayerEntity =
		FabricPlayerEntity( backingEntity as McPlayerEntity )

	override fun isChild(): Boolean =
		backingEntity.isBaby

	override fun isWaterMob(): Boolean =
		backingEntity.canBreatheInWater()

	override fun isClimbing(): Boolean =
		backingEntity.isClimbing

	override fun getArmorProtection(): Int =
		backingEntity.armor

	override fun getItemInMainHand(): ItemStack =
		FabricItemStack( backingEntity.getStackInHand( McHand.MAIN_HAND ) )

	override fun getItemInOffHand(): ItemStack =
		FabricItemStack( backingEntity.getStackInHand( McHand.OFF_HAND ) )

	override fun hasEquipment( slot: EquipmentSlot ): Boolean =
		backingEntity.hasStackEquipped( slot.toMC() )

	override fun getEquipment( slot: EquipmentSlot ): ItemStack =
		FabricItemStack( backingEntity.getEquippedStack( slot.toMC() ) )

	override fun setEquipment( slot: EquipmentSlot, stack: ItemStack ) {
		backingEntity.equipStack( slot.toMC(), stack.toMC() )
	}
}

